src/mcts/search.cc

```diff
@@ -496,31 +523,52 @@ std::vector<EdgeAndNode> Search::GetBestChildrenNoTemperature(Node* parent,
     PopulateRootMoveLimit(&root_limit);
   }
   // Best child is selected using the following criteria:
+  // with Certainty Propagation:
+  // * Prefer certain wins, avoid certain losses
+  // Otherwise:
   // * Largest number of playouts.
   // * If two nodes have equal number:
   //   * If that number is 0, the one with larger prior wins.
   //   * If that number is larger than 0, the one with larger eval wins.
-  using El = std::tuple<uint64_t, float, float, EdgeAndNode>;
+  using El = std::tuple<int, uint64_t, float, float, EdgeAndNode>;
   std::vector<El> edges;
   for (auto edge : parent->Edges()) {
     if (parent == root_node_ && !root_limit.empty() &&
         std::find(root_limit.begin(), root_limit.end(), edge.GetMove()) ==
             root_limit.end()) {
       continue;
     }
-    edges.emplace_back(edge.GetN(), edge.GetQ(0), edge.GetP(), edge);
+    edges.emplace_back((params_.GetCertaintyPropagation())
+                           ? edge.edge()->GetEQ() * (edge.IsTerminal()+1)
+            : 0,
+        edge.GetN(), edge.GetQ(0), edge.GetP(), edge);
   }
+  // In case of certain draws, insert
+  // these draws at first N (descending) where Q<=0.
+  if (params_.GetCertaintyPropagation()) {
+    std::partial_sort(edges.begin(), edges.end(), edges.end(),
+                      std::greater<El>());
+    // largest N with Q >= 0
+    uint64_t largest_N = 0;
+    for (auto it = edges.begin(); it != edges.end(); ++it) {
+      if (std::get<2>(*it) <= 0.0f && largest_N == 0)
+        largest_N = std::get<1>(*it);
+      if (std::get<4>(*it).edge()->IsCertainDraw() && largest_N > 0)
```

Comment:
I had the double loop in PR487 at one point in time. I convinced myself that this was not necessary and simplified the code. Your example is actually the desired behaviour. We want to choose by visits, except if a high visit move has a Q<0 then we prefer the certain draw. Intuitively, we take a well explored draw (certain or not) over a not well explored higher Q. In this regard our decision rule does not change with CP. But we prefer a certain draw over any Q < 0.00 move. 

The rational mainly stems from pruning certain draws altogether then you have to reinsert them. This was a function of PR487 at one point in time. I will go over this again, tomorrow (when i am more awake) and see whether (a) it really works as intended and (b) if it is necessary at all, as i currently do not prune in PR700. 

Thanks very much for going over this, as this part has changed often over time and while i think its ok, one never knows with all the refactoring going on.

Reply:
Thanks for the explanation.

Maybe it's a good idea to update the comment to say this is intentional? Right now it says "In case of certain draws, insert these draws at first N (descending) where Q<=0", but what is actually meant (what the code is doing) is "Ensure that certain draws have at least as many N as the first N (descending) where Q<=0" or something along those lines.

This will save some time of future code reviewers that think they've found a bug.

Actually, now that I think of it, shouldn't you put those nodes at N+1 (assume N of the first node with Q<=0 is actually very low Q<<0, then it might still be ordered at the end in front of our certain draw)...

Reply:
Yes, i will absolutely change the comment the next time i update this PR.

Reply:
comment changed, please check if new wording is clear.

Reply:
Thanks, ok for me!